src/wp-includes/block-template-utils.php

```diff
@@ -1444,36 +1444,69 @@ function get_template_hierarchy( $slug, $is_custom = false, $template_prefix = '
  * @since 6.5.0
  * @access private
  *
- * @param stdClass        $post    An object representing a template or template part
+ * @param stdClass        $changes An object representing a template or template part
  *                                 prepared for inserting or updating the database.
  * @param WP_REST_Request $request Request object.
  * @return stdClass The updated object representing a template or template part.
  */
-function inject_ignored_hooked_blocks_metadata_attributes( $post, $request ) {
+function inject_ignored_hooked_blocks_metadata_attributes( $changes, $request ) {
 	$filter_name = current_filter();
 	if ( ! str_starts_with( $filter_name, 'rest_pre_insert_' ) ) {
-		return $post;
+		return $changes;
 	}
 	$post_type = str_replace( 'rest_pre_insert_', '', $filter_name );
 
 	$hooked_blocks = get_hooked_blocks();
 	if ( empty( $hooked_blocks ) && ! has_filter( 'hooked_block_types' ) ) {
-		return $post;
+		return $changes;
 	}
 
-	// At this point, the post has already been created.
 	// We need to build the corresponding `WP_Block_Template` object as context argument for the visitor.
 	// To that end, we need to suppress hooked blocks from getting inserted into the template.
 	add_filter( 'hooked_block_types', '__return_empty_array', 99999, 0 );
-	$template = $request['id'] ? get_block_template( $request['id'], $post_type ) : null;
+
+	// We also need to mimic terms and meta for the post based on the corresponding
+	// `terms_input` and `meta_input` properties in the changes object.
+
+	$terms_filter = function( $terms, $post_id, $taxonomy ) use ( $changes ) {
+		if ( 'wp_theme' !== $taxonomy || ! isset( $changes->tax_input['wp_theme'] ) ) {
+			return $terms;
+		}
+
+		// TODO: Verify it's not an error object.
+		// TODO: Verify that $post_id matches (or isn't set).
+		$term       = new stdClass;
+		$term->name = $changes->tax_input['wp_theme'];
+
+		$term = new WP_Term( $term );
+		return array( $term );
+	};
+
+	$meta_filter = function( $value, $post_id, $meta_key, $single ) use ( $changes ) {
+		if ( 'origin' === $meta_key && isset( $changes->meta_input['origin'] ) ) {
+			return $single ? $changes->meta_input['origin'] : array( $changes->meta_input['origin'] );
+		}
+
+		if ( 'is_wp_suggestion' === $meta_key && isset( $changes->meta_input['is_wp_suggestion'] ) ) {
+			return $single ? $changes->meta_input['is_wp_suggestion'] : array( $changes->meta_input['is_wp_suggestion'] );
+		}
+
+		return $value;
+	};
+
+	add_filter( 'get_the_terms', $terms_filter, 10, 3 );
+	add_filter( 'get_post_metadata', $meta_filter, 10, 4 );
```

Comment:
This is a bit too wild, and I don’t think I want to land it like this. However, it has helped inform a strategy in that it has made clear that a `WP_Block_Template` needs more information than is solely contained in a `WP_Post` (which it fetches from meta and terms). This becomes a problem when we try to pass `$changes` to `_build_block_template_result_from_post`, as it doesn’t have that meta and terms in the DB available yet; instead, it uses the `meta_input` and `tax_input` fields to set those upon insertion into the DB.

Let’s extract a helper function from `_build_block_template_result_from_post` that acknowledges this and accepts as its arguments a `WP_Post`, and additionally, all the required information that’s not included in it, i.e. a `$theme`, an `$origin`, an `$is_wp_suggestion`, etc. That way, we’ll have a function that accurately represents the mapping from `WP_Post` object to `WP_Block_Template`; and in our `inject_ignored_hooked_blocks_metadata_attributes`, we can provide the additional required arguments by reading them from `meta_input` and `tax_input`.